### DISCUSSION TITLE
Add non inline versions of data/add/set/update methods to firestore

### DISCUSSION THREAD

**Discussion by mr-kew** [GitLiveApp/firebase-kotlin-sdk#439] [Ideas]
| Library               | Class                            | Member                                        |
| :-----------------| :------------------------ | :----------------------------------- |
| firestore | DocumentSnapshot | data  |
| firestore | CollectionReference | add  |
| firestore | DocumentReference | set  |
| firestore | DocumentReference | update  |

These methods have only `inline fun <reified T>` versions of them which severely limits their usability in kotlin code since you either have to call them with a concrete type or you have to propagate `inline` up into your code (which then cannot be virtual, etc).

I think they should have a `KClass<T>` version of them so they can be used with generics in kotlin. 

This is how it could look like:

    fun <T> DocumentSnapshot.data(clazz: KClass<out T>): T {
        ...
    }

I don't really know if it is even possible, but from user's perspective it sure is needed. I opened it first as an disscussion in case it is a complete nonsense.

**Comment by nbransby** [comment]
There should be versions of those methods that take a serializer - that's the one you want to use if you want to avoid the reified type

**Reply by mr-kew** [reply]
Oh, how did I miss it? Thats exactly what I was looking for! Thank you very much!

Just in case anyone finds this, you can achieve it this way:

    fun <T> myFunc(clazz: KClass<T>, snapshot: DocumentSnapshot) = snapshot.data(strategy = clazz.serializer())
